### DISCUSSION TITLE
Multi Facet Navigation issue with NEXT JS 14 App Router.

### DISCUSSION THREAD

**Discussion by Jbmanllr** [vercel/next.js#60475] [App Router]
Hello everyone,

I'm currently facing a challenge with a Next.js 14 project using APP Router and would appreciate some insights or suggestions from the community.

**Primary Issue**: I'm struggling to find a method to both acquire necessary information for constructing an API URL and utilize this URL for server-side data fetching, specifically for components that are located within the layout but outside the main page.tsx.

Before i explain more in depth here is my app structure and simplified code:

&#x200B;

* page.tsx
* navigation.tsx
* layout.tsx
* context.tsx
* Categories
* \[...slug\]
* page.tsx

LISTING CONTEXT.tsx (USE CLIENT)

    'use client';
    const ListingContext = createContext(undefined);
    
    export const ListingProvider = ({
      children,
    }) => {
    
      // Using usePathname() in context
      const path = usePathname();
      const searchParams = useSearchParams();
    
      const [pathName, setPathName] = useState(path);
      const [searchParams, setSearchParams] = useState(path);
      const [selectedFacets, setSelectedFacets] = useState({});
    
      React.useEffect(() => {
        setPathName(path);
        setSearchParams(searchParams);
      }, [path, searchParams]);
    
      const setFacetState = (facetName: string, newSelectedItems: any) => {
        setSelectedFacets((prev) => ({
          ...prev,
          [facetName]: newSelectedItems,
        }));
      };
    
      return (
        <ListingContext.Provider
          value={{
            selectedFacets,
            setFacetState,
            pathName,
            searchParams
          }}
        >
          {children}
        </ListingContext.Provider>
      );
    };
    
    export const useListing = () => {
      const context = useContext(ListingContext);
      if (context === undefined) {
        throw new Error('useListing must be used within a ListingProvider');
      }
      return context;
    };

LISTING LAYOUT.tsx (USE SERVER)

    export default async function ListingLayout(props) {
    
      return (
        <ListingProvider>
          <Navigation/>
          {props.children}
        </ListingProvider>
      );
    }

LISTING PAGE.tsx (USE SERVER)

    async function Items({params, searchParams}) {
      
      const apiUrl = {*** BUILD API URL WITH HELP OF PARAMS AND SEARCH PARAMS ***}
    
      // FETCH SERVER SIDE DATA
      let data = await fetch(ApiUrl);
      let items = await data.json();
    
      return (
        <div>
          {items.data.map((item, index) => (
           {*** LIST LAYOUT ***}
          ))}
        </div>
      );
    }
    
    export default function Page({
      searchParams,
      params,
    }: {
      params: { slug: string };
      searchParams: { [key: string]: string | string[] | undefined };
    }) {
      return (
          <div>  
              <Suspense fallback={<p>Loading...</p>}>
                  <Items params={params} searchParams={searchParams} />
              </Suspense>
          </div>
      );
    }

LISTING NAVIGATION.tsx (USE CLIENT)

    'use client';
    
    export function Navigation(props) {
    
    ***It's a client component, here i can use hooks and my context data, i could directly call useParams or useSearchParams, but i did in in my context so i'm using my context.***
    
      const { pathName, searchParams, selectedFacets, setFacetState } = useListing();
      const [navdata, setNavdata] = React.useState(null);
    
      const getClientData = async () => {
    
        const apiUrl = {*** BUILD API URL WITH HELP OF PARAMS AND SEARCH PARAMS ***}
    
        try {
          const response = await fetch(
            apiUrl,
            {
              method: 'GET',
            },
          );
          if (response) {
            const data = await response.json();
            setNavdata(data);
          }
        } catch (error) {
          console.log(error);
        }
      };
      
      
      React.useEffect(() => {
        getClientData();
      }, [pathName, searchParams]);
    
    
      return (
        <nav>
          <ComboBox
            options={navdata}
          />
        </nav>
      );
    }

The core of my challenge lies in accessing searchParams and params, which are only available on the server side within page.tsx. Utilizing hooks like useSearchParams and useParams transforms my component into a client-side component, consequently preventing me from fetching data on the server side.

Currently, the setup works fine when fetching navigation data on the client side (as seen in navigation.tsx). However, my requirement is to fetch this data on the server side. The catch is that I can only access params and searchParams server-side within page.tsx.

To reiterate: My objective is to FETCH THE NAVIGATION DATA ON THE SERVER SIDE. The current implementation, where navigation data is fetched on the client side, is merely a temporary solution. The main issue I'm trying to resolve is finding a method to retrieve proper navigation data on the server side.

So the solution would either to: be able to get the params and searchParams on the server inside a server component that is outside of page.tsx. Fetch data server side inside a client component.

Am i missing something, do i have an architecture problem?

**Comment by hedysnike** [comment]
```
export default async function IndexPage({ params: { lang, route }, searchParams }: { params: { lang: Locale; route: string }; searchParams: any }) {
```

you DO can you use params on the server but instead you should basically declare it on page and pass it down to the components
below is example on how i'm retriving the data from params

so url localhost:3000/en/somepage?id=SOMEID
```
  const id = searchParams?.id;
```
id would be SOMEID 
on the server

**Reply by hedysnike** [reply]
also, there's is catch

when using searchparams on the server
it will make your page dynamic instead of static so, everything will be slower basically
it might be actually better to leave it on client side

**Comment by vicentel89** [comment]
You cannot fetch data on the server from a client component. The way to fetch data on the server is using fetch in a server component. Then you can pass that data through props to other sever or client components.

You can change search parameters from a client component. When a page notice that the search parameters changed, it will rerender and refetch the data.

It is supposed that using fetch in nextjs would help you to avoid context because it caches the data.

I created something similar where I am changing the search params from the client and fetching data on the server.

Take a look.

Here I change the search params using replace from useRouter.

[https://github.com/vicentel89/marvel-characters/blob/main/src/app/(home)/_components/search.tsx](https://github.com/vicentel89/marvel-characters/blob/main/src/app/(home)/page.tsx)


But the data is gotten in a server component.

[https://github.com/vicentel89/marvel-characters/blob/main/src/app/(home)/page.tsx](https://github.com/vicentel89/marvel-characters/blob/main/src/app/(home)/page.tsx)
